### ISSUE TITLE
ban use of types that inherit from a specific parent

### ISSUE THREAD

**Issue by fommil** [scalacenter/scalafix#531]
e.g. I would like to ban use of types that extend from `Throwable` in public apis.

**Comment by vovapolu** [comment]
According to what @olafurpg says, scalafix isn't able to determine such relationships between types for now.

**Comment by olafurpg** [comment]
For cases like Throwable, it might be doable with java runtime reflection, something like
```scala
@ def toFqn(symbol: Symbol) = symbol.syntax.replaceAll("\\.$", "\\$").stripSuffix("#").stripPrefix("_root_.")
defined function toFqn

@ val cl = this.getClass.getClassLoader
cl: ClassLoader = ammonite.runtime.SpecialClassLoader@214b199c

// companion object
@ classOf[Throwable].isAssignableFrom(cl.loadClass(toFqn(Symbol("_root_.scala.reflect.io.FileOperationException."))))
res33: Boolean = false

// class
@ classOf[Throwable].isAssignableFrom(cl.loadClass(toFqn(Symbol("_root_.scala.reflect.io.FileOperationException#"))))
res34: Boolean = true
```

This won't work with higher kinded types. scalafix-core also cross-builds to Scala.js where this probably won't work either, but as long as the rule is in another jvm-only module or the utility methods throw UnsupportedException in JS then you can try to use this.

**Comment by olafurpg** [comment]
The `toFqn` method is hacky, Symbol is a small ADT that can be pattern matched on for a more robust translation. For example, local symbols don't have fully qualified names so they need special handling.

**Comment by fommil** [comment]
this is related to #597 but perhaps not the same

**Comment by olafurpg** [comment]
This is a good idea. After the PR https://github.com/scalacenter/scalafix/pull/827, we are moving most built-in rules outside of the scalacenter/scalafix repository so that they can evolve at a different pace than Scalafix APIs. As long as a custom rule meets a certain level of quality, it can still 

- be automatically be included by sbt-scalafix
- have tab completion in the sbt shell like built-in rules
- be documented on the main Scalafix website

Ping us on Gitter if you believe your rule is good enough for inclusion. Don't hesitate to ask for help around code review or general guidance on writing/sharing custom rules.

**Comment by ghostbuster91** [comment]
Since there is no option to check these type relations in scalafix I tried using provided `toFqn` method hack. However, in my case I am dealing with `RuntimeException` that comes from `scala.package` and I need to desugar it before passing to the classLoader. I followed example of desugaring here: https://github.com/scalacenter/scalafix/blob/main/docs/developers/semantic-type.md#dealias-types but it is also a hack and in my case it returns just `RuntimeException` which is incorrect. 

Since the proper handling of type aliases was apparently merged into semantic-db 3.3.0 (https://github.com/scalameta/scalameta/issues/863#issuecomment-367054185) can we make that information available in scalafix?
